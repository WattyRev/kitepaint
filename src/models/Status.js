import PropTypes from "prop-types";

/**
 * Status can be applied to any model to indicate it's public visibility.
 */
export const statusProp = PropTypes.oneOf(["0", "1", "2"]);
const Status = {
  "0": "Private",
  "1": "Unlisted",
  "2": "Public",
  PRIVATE: "0",
  UNLISTED: "1",
  PUBLIC: "2"
};
export default Status;
