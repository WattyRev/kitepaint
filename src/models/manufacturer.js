import PropTypes from "prop-types";

/**
 * A manufacturer is a company that manufactures kites and owns one or more products.
 */
const manufacturerState = PropTypes.shape({
  id: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  logo: PropTypes.string.isRequired,
  website: PropTypes.string
});

export default manufacturerState;

/**
 * A mock manufacturer used for testing
 * @type {Object}
 */
const getMockManufacturer = () => ({
  id: "abc",
  name: "Krazy Kites",
  logo: "abc.jpg",
  website: "http://krazykites.com"
});
export { getMockManufacturer };
