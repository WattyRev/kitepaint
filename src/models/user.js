import PropTypes from "prop-types";

/**
 * A user is an individual user account. It contains limited details that are available publically.
 */
const userShape = PropTypes.shape({
  username: PropTypes.string.isRequired,
  loginid: PropTypes.string.isRequired
});

export default userShape;

/**
 * A mock user used for testing.
 */
const getMockUser = () => ({
  username: "FlyFucker",
  loginid: "abc"
});

export { getMockUser };
