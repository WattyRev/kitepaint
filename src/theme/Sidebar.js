import React from "react";
import styled, { css } from "styled-components";
import PropTypes from "prop-types";
import BlockListItem from "./BlockListItem";

/**
 * An overall wrapper for the sidebar
 */
export const StyleWrapper = styled.div`
  width: 175px;
  display: block;
  flex-shrink: 0;
  background: ${props => props.theme.colors.grayDarker};
  box-shadow: 0 0 2px 2px ${props => props.theme.colors.black};
  position: relative;
`;

/**
 * An individual, possibly selectable, item that appreas in the sidebar.
 *
 * @param {Boolean} hasAction indicates that the item is actionable. Adds a hover state
 * @param {Boolean} isActive indicates that the item is active.
 */
export const Item = styled(BlockListItem)`
  display: flex;
  align-items: center;
  cursor: ${props => (props.hasAction ? "pointer" : "default")};
  transition: 0.3s background;
  position: relative;

  &:after {
    content: "";
    display: block;
    position: absolute;
    height: 100%;
    width: ${props => (props.isActive ? 8 : 0)}px;
    background: ${props => props.theme.colors.silver};
    top: 0;
    right: 0;
    transition: 0.3s width;
  }
  ${props =>
    props.hasAction
      ? css`
          &:hover {
            background: ${props.theme.colors.black};
          }
        `
      : null};

  > * {
    margin: 0 4px;
  }
`;

/**
 * A heading that can appear above a group of ListItems.
 */
const Heading = styled(BlockListItem)`
  padding: 16px 8px 4px;
  font-weight: bold;
  font-size: 12px;
  white-space: initial;
`;

/**
 * The sidebar displayed when editing/creating a new design
 */
const Sidebar = ({ children }) => (
  <StyleWrapper>
    {children({
      components: {
        Item,
        Heading
      }
    })}
  </StyleWrapper>
);

Sidebar.propTypes = {
  children: PropTypes.func.isRequired
};

export default Sidebar;
