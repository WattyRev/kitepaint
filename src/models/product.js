import PropTypes from "prop-types";
import { getMockManufacturer } from "./manufacturer";
import Status, { statusProp } from "./status";

/**
 * A product is a kite that can be customized to create a design.
 */
const productShape = PropTypes.shape({
  id: PropTypes.string.isRequired,
  embed: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  notes: PropTypes.arrayOf(PropTypes.string),
  manufacturer: PropTypes.string.isRequired,
  status: statusProp.isRequired,
  url: PropTypes.string,
  colors: PropTypes.arrayOf(
    PropTypes.shape({
      name: PropTypes.string.isRequired,
      color: PropTypes.string.isRequired
    })
  ),
  variations: PropTypes.arrayOf(
    PropTypes.shape({
      name: PropTypes.string.isRequired,
      svg: PropTypes.string.isRequired
    })
  )
});

export default productShape;

/**
 * A mock product used for testing
 * @type {Object}
 */
const getMockProduct = () => ({
  id: "product-1",
  embed: "krazykites.com",
  name: "Krazy Kite",
  manufacturer: getMockManufacturer().id,
  url: "http://krazykites.com/krazy-kite",
  colors: [
    {
      name: "red",
      color: "#ff0000"
    },
    {
      name: "black",
      color: "#000000"
    }
  ],
  variations: [
    {
      name: "Standard",
      svg: "<div>Kool Kite</div>"
    }
  ],
  status: Status.PUBLIC
});

export { getMockProduct };
