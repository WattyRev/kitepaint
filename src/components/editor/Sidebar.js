import React from "react";
import styled, { css } from "styled-components";
import PropTypes from "prop-types";
import manufacturerShape from "../../models/manufacturer";
import productShape from "../../models/product";
import designShape from "../../models/design";
import { BlockListItem, Icon } from "../../theme";
import ManufacturerLogo from "../ManufacturerLogo";
import ColorTile from "./ColorTile";
import ColorableSvg from "./ColorableSvg";

/**
 * An overall wrapper for the sidebar
 */
export const StyleWrapper = styled.div`
  width: 175px;
  display: block;
  flex-shrink: 0;
  background: ${props => props.theme.colors.grayDarker};
  box-shadow: 0 0 2px 2px ${props => props.theme.colors.black};
`;

/**
 * An individual, possibly selectable, item that appreas in the sidebar.
 */
export const ListItem = styled(BlockListItem)`
  display: flex;
  align-items: center;
  cursor: ${props => (props.hasAction ? "pointer" : "default")};
  transition: 0.3s background;
  position: relative;

  &:after {
    content: "";
    display: block;
    position: absolute;
    height: 100%;
    width: ${props => (props.isActive ? 8 : 0)}px;
    background: ${props => props.theme.colors.silver};
    top: 0;
    right: 0;
    transition: 0.3s width;
  }
  ${props =>
    props.hasAction
      ? css`
          &:hover {
            background: ${props.theme.colors.black};
          }
        `
      : null};

  > * {
    margin: 0 4px;
  }
`;

/**
 * A heading that can appear above a group of ListItems.
 */
const ListHeading = styled(BlockListItem)`
  padding: 16px 8px 2px;
  font-weight: bold;
  font-size: 12px;
`;

/**
 * A styled preview of a variation's svg
 */
const VariationPreview = styled.div`
  display: block;
  width: 24px;
  height: 24px;

  svg {
    margin-top: 50%;
    transform: translateY(-50%);
  }
`;

/**
 * The sidebar displayed when editing/creating a new design
 */
const Sidebar = ({
  appliedColors,
  manufacturer,
  product,
  design,
  selectedVariation,
  selectedColor,
  onColorSelect,
  onVariationSelect
}) => (
  <StyleWrapper>
    <ListItem
      className="testing_manufacturer"
      isLight
      as={product.url || manufacturer.website ? "a" : "div"}
      href={product.url || manufacturer.website}
      target="_blank"
      hasAction={!!(product.url || manufacturer.website)}
    >
      <ManufacturerLogo
        size={32}
        noMargin
        src={`/logos/${manufacturer.logo}`}
      />
      <div>
        {product.name}
        <br />
        <small>by {manufacturer.name}</small>
      </div>
    </ListItem>
    {design && (
      <ListHeading className="testing_design" isLight>
        {design.name}
      </ListHeading>
    )}
    {product.variations.map(variation => (
      <ListItem
        className="testing_variation"
        isLight
        hasAction
        isActive={variation.name === selectedVariation}
        key={variation.name}
        onClick={() => onVariationSelect(variation.name)}
      >
        <VariationPreview>
          <ColorableSvg
            svg={variation.svg}
            colorMap={appliedColors[variation.name] || {}}
          />
        </VariationPreview>{" "}
        {variation.name}
      </ListItem>
    ))}
    <ListHeading isLight>
      <Icon icon="palette" /> Colors
    </ListHeading>
    {product.colors.map(color => (
      <ListItem
        className="testing_color"
        isLight
        hasAction
        key={color.name}
        isActive={color.name === selectedColor}
        onClick={() => onColorSelect(color.name)}
      >
        <ColorTile color={color.color} /> {color.name}
      </ListItem>
    ))}
  </StyleWrapper>
);

Sidebar.propTypes = {
  appliedColors: PropTypes.objectOf(
    PropTypes.objectOf(
      PropTypes.shape({
        name: PropTypes.string.isRequired,
        color: PropTypes.string.isRequired
      })
    )
  ),
  /**
   * The manufacturer that creates the product being edited.
   */
  manufacturer: manufacturerShape.isRequired,
  /**
   * The produt that is being edited
   */
  product: productShape.isRequired,
  /**
   * The pre-existing design that is being edited, if any
   */
  design: designShape,
  /**
   * The name of the variation that is currently selected
   */
  selectedVariation: PropTypes.string.isRequired,
  /**
   * The name of the color that is currently selected
   */
  selectedColor: PropTypes.string.isRequired,
  /**
   * Called when a color is selected.
   * Is provided with the color name as the first parameter.
   */
  onColorSelect: PropTypes.func.isRequired,
  /**
   * Called when a variation is selected.
   * Is provided with the variation name as the first parameter.
   */
  onVariationSelect: PropTypes.func.isRequired
};

export default Sidebar;
