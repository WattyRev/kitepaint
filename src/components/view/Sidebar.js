import React from "react";
import styled from "styled-components";
import PropTypes from "prop-types";
import manufacturerShape from "../../models/manufacturer";
import productShape from "../../models/product";
import designShape from "../../models/design";
import userShape from "../../models/user";
import ColorTile from "../editor/ColorTile";
import { Icon, FillToBottom, Sidebar as SidebarUI } from "../../theme";
import ManufacturerLogo from "../ManufacturerLogo";
import Svg from "../Svg";

/**
 * A styled preview of a variation's svg
 */
const VariationPreview = styled.div`
  display: block;
  width: 24px;
  height: 24px;

  svg {
    margin-top: 50%;
    transform: translateY(-50%);
  }
`;

/**
 * The sidebar displayed when editing/creating a new design
 */
const Sidebar = ({
  manufacturer,
  product,
  design,
  selectedVariation,
  onVariationSelect,
  user,
  usedColors
}) => (
  <SidebarUI>
    {sidebar => (
      <React.Fragment>
        {product &&
          manufacturer && (
            <sidebar.components.Item
              isLight
              as={product.url || manufacturer.website ? "a" : "div"}
              href={product.url || manufacturer.website}
              target="_blank"
              hasAction={!!(product.url || manufacturer.website)}
            >
              <ManufacturerLogo
                size={32}
                noMargin
                src={`/logos/${manufacturer.logo}`}
              />
              <div className="testing_manufacturer">
                {product.name} <br />
                <small>by {manufacturer.name}</small>
              </div>
            </sidebar.components.Item>
          )}
        <sidebar.components.Heading className="testing_design" isLight>
          {design.name}{" "}
          {user && (
            <React.Fragment>
              <br />
              designed by {user.username}
            </React.Fragment>
          )}
        </sidebar.components.Heading>
        {design.variations.map(variation => (
          <sidebar.components.Item
            className="testing_variation"
            isLight
            hasAction
            isActive={variation.name === selectedVariation}
            key={variation.name}
            onClick={() => onVariationSelect(variation.name)}
          >
            <VariationPreview>
              <Svg svg={variation.svg} />
            </VariationPreview>{" "}
            {variation.name}
          </sidebar.components.Item>
        ))}
        <sidebar.components.Heading isLight>
          <Icon icon="palette" /> Colors Used
        </sidebar.components.Heading>
        <FillToBottom offset={35}>
          {usedColors[selectedVariation].map(color => (
            <sidebar.components.Item isLight key={color.color}>
              <ColorTile color={color.color} /> {color.name || color.color}
            </sidebar.components.Item>
          ))}
        </FillToBottom>
      </React.Fragment>
    )}
  </SidebarUI>
);

Sidebar.propTypes = {
  /**
   * The manufacturer that creates the product being edited.
   */
  manufacturer: manufacturerShape,
  /**
   * The produt that is being edited
   */
  product: productShape,
  /**
   * The pre-existing design that is being edited, if any
   */
  design: designShape.isRequired,
  /**
   * The user that created the design.
   */
  user: userShape,
  /**
   * The name of the variation that is currently selected
   */
  selectedVariation: PropTypes.string.isRequired,
  /**
   * The colors that are used in the current variation.
   */
  usedColors: PropTypes.objectOf(
    PropTypes.arrayOf(
      PropTypes.shape({
        name: PropTypes.string,
        color: PropTypes.string.isRequired
      }).isRequired
    ).isRequired
  ).isRequired,
  /**
   * Called when a variation is selected.
   * Is provided with the variation name as the first parameter.
   */
  onVariationSelect: PropTypes.func.isRequired
};

export default Sidebar;
