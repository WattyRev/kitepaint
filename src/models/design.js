import PropTypes from "prop-types";
import { getMockProduct } from "./product";
import Status, { statusProp } from "./status";

/**
 * A design is a user-created entity. It represetns a saved design of a particular product.
 */
const designShape = PropTypes.shape({
  id: PropTypes.string.isRequired,
  // A string representing the created date
  created: PropTypes.string.isRequired,
  // A string representing the updated date
  updated: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  // The ID of the cooresponding product
  product: PropTypes.string.isRequired,
  // The ID of the user that created the design
  user: PropTypes.string,
  variations: PropTypes.arrayOf(
    PropTypes.shape({
      name: PropTypes.string.isRequired,
      svg: PropTypes.string.isRequired,
      primary: PropTypes.bool.isRequired
    })
  ),
  // The status of the cooresponding product
  productStatus: statusProp.isRequired,
  // The status of the design
  status: statusProp.isRequired
});

export default designShape;

/**
 * A mock design used for testing
 * @type {Object}
 */
const getMockDesign = () => ({
  id: "design-1",
  created: "01/01/2018",
  updated: "01/02/2018",
  name: "picasso",
  product: getMockProduct().id,
  variations: [
    {
      name: "Standard",
      svg: "<div>picasso standard</div>",
      primary: true
    }
  ],
  status: Status.PRIVATE,
  productStatus: Status.PUBLIC
});
export { getMockDesign };
