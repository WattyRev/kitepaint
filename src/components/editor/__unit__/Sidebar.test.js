import React from "react";
import { shallow } from "enzyme";
import { getMockManufacturer } from "../../../models/manufacturer";
import { getMockProduct } from "../../../models/product";
import { getMockDesign } from "../../../models/Design";
import Sidebar, { StyledSidebar } from "../Sidebar";

describe("Sidebar", () => {
  let defaultProps;
  let sidebarUIData;
  beforeEach(() => {
    defaultProps = {
      appliedColors: {},
      manufacturer: getMockManufacturer(),
      product: getMockProduct(),
      selectedVariation: "Standard",
      selectedColor: "red",
      onColorSelect: jest.fn(),
      onVariationSelect: jest.fn()
    };

    sidebarUIData = {
      components: {
        Item: ({ children }) => children,
        Heading: ({ children }) => children
      }
    };
  });
  it("renders", () => {
    const wrapper = shallow(<Sidebar {...defaultProps} />);
    shallow(
      <div>{wrapper.find(StyledSidebar).prop("children")(sidebarUIData)}</div>
    );
  });
  it("makes the manufacturer list item a link if the product has a url", () => {
    expect.assertions(2);
    defaultProps.product.url = "http://zombo.com";
    const wrapper = shallow(<Sidebar {...defaultProps} />);
    const sidebarContent = shallow(
      <div>{wrapper.find(StyledSidebar).prop("children")(sidebarUIData)}</div>
    );
    expect(sidebarContent.find(".testing_manufacturer").prop("as")).toEqual(
      "a"
    );
    expect(sidebarContent.find(".testing_manufacturer").prop("href")).toEqual(
      "http://zombo.com"
    );
  });
  it("makes the manufacturer list item a link if the product does not have a url, but the manufacturer has a website", () => {
    expect.assertions(2);
    defaultProps.product.url = null;
    defaultProps.manufacturer.website = "http://albinoblacksheep.com";
    const wrapper = shallow(<Sidebar {...defaultProps} />);
    const sidebarContent = shallow(
      <div>{wrapper.find(StyledSidebar).prop("children")(sidebarUIData)}</div>
    );
    expect(sidebarContent.find(".testing_manufacturer").prop("as")).toEqual(
      "a"
    );
    expect(sidebarContent.find(".testing_manufacturer").prop("href")).toEqual(
      "http://albinoblacksheep.com"
    );
  });
  it("makes the manufacturer list item a div if there is no url or website", () => {
    expect.assertions(1);
    defaultProps.product.url = null;
    defaultProps.manufacturer.website = null;
    const wrapper = shallow(<Sidebar {...defaultProps} />);
    const sidebarContent = shallow(
      <div>{wrapper.find(StyledSidebar).prop("children")(sidebarUIData)}</div>
    );
    expect(sidebarContent.find(".testing_manufacturer").prop("as")).toEqual(
      "div"
    );
  });
  it("displays a heading with the design name if there is a design", () => {
    expect.assertions(2);
    defaultProps.design = getMockDesign({
      name: "Boogers"
    });
    const wrapper = shallow(<Sidebar {...defaultProps} />);
    const sidebarContent = shallow(
      <div>{wrapper.find(StyledSidebar).prop("children")(sidebarUIData)}</div>
    );
    expect(sidebarContent.find(".testing_design")).toHaveLength(1);
    expect(
      sidebarContent
        .find(".testing_design")
        .prop("children")
        .trim()
    ).toEqual("Boogers");
  });
  it("does not display a heading with the design name if there is no design", () => {
    expect.assertions(1);
    defaultProps.design = null;
    const wrapper = shallow(<Sidebar {...defaultProps} />);
    const sidebarContent = shallow(
      <div>{wrapper.find(StyledSidebar).prop("children")(sidebarUIData)}</div>
    );
    expect(sidebarContent.find(".testing_design")).toHaveLength(0);
  });
  it("triggers onVariationSelect when a variation is selected", () => {
    expect.assertions(2);
    defaultProps.product.variations = [
      {
        name: "Standard",
        svg: "<div>Kool Kite</div>"
      },
      {
        name: "Vented",
        svg: "<div>Kool Vented Kite</div>"
      }
    ];
    const wrapper = shallow(<Sidebar {...defaultProps} />);
    const sidebarContent = shallow(
      <div>{wrapper.find(StyledSidebar).prop("children")(sidebarUIData)}</div>
    );
    sidebarContent
      .find(".testing_variation")
      .at(1)
      .simulate("click");
    expect(defaultProps.onVariationSelect).toHaveBeenCalled();
    expect(defaultProps.onVariationSelect.mock.calls[0][0]).toEqual("Vented");
  });
  it("triggers onColorSelect when a color is selected", () => {
    expect.assertions(2);
    defaultProps.product.colors = [
      {
        name: "red",
        color: "#ff0000"
      },
      {
        name: "black",
        color: "#000000"
      }
    ];
    const wrapper = shallow(<Sidebar {...defaultProps} />);
    const sidebarContent = shallow(
      <div>{wrapper.find(StyledSidebar).prop("children")(sidebarUIData)}</div>
    );
    sidebarContent
      .find(".testing_color")
      .at(1)
      .simulate("click");
    expect(defaultProps.onColorSelect).toHaveBeenCalled();
    expect(defaultProps.onColorSelect.mock.calls[0][0]).toEqual("black");
  });
});
