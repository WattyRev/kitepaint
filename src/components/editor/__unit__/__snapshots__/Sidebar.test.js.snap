// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sidebar renders 1`] = `
<styled.div>
  <Styled(styled.div)
    as="a"
    className="testing_manufacturer"
    hasAction={true}
    href="http://krazykites.com/krazy-kite"
    isLight={true}
    target="_blank"
  >
    <styled.div
      noMargin={true}
      size={32}
      src="/logos/abc.jpg"
    />
    <div>
      Krazy Kite
      <br />
      <small>
        by 
        Krazy Kites
      </small>
    </div>
  </Styled(styled.div)>
  <Styled(styled.div)
    className="testing_variation"
    hasAction={true}
    isActive={true}
    isLight={true}
    key="Standard"
    onClick={[Function]}
  >
    <styled.div>
      <ColorableSvg
        colorMap={Object {}}
        svg="<div>Kool Kite</div>"
      />
    </styled.div>
     
    Standard
  </Styled(styled.div)>
  <Styled(styled.div)
    isLight={true}
  >
    <FontAwesomeIcon
      border={false}
      className=""
      fixedWidth={false}
      flip={null}
      icon="palette"
      inverse={false}
      listItem={false}
      mask={null}
      pull={null}
      pulse={false}
      rotation={null}
      size={null}
      spin={false}
      symbol={false}
      title=""
      transform={null}
    />
     Colors
  </Styled(styled.div)>
  <Styled(styled.div)
    className="testing_color"
    hasAction={true}
    isActive={true}
    isLight={true}
    key="red"
    onClick={[Function]}
  >
    <styled.div
      color="#ff0000"
    />
     
    red
  </Styled(styled.div)>
  <Styled(styled.div)
    className="testing_color"
    hasAction={true}
    isActive={false}
    isLight={true}
    key="black"
    onClick={[Function]}
  >
    <styled.div
      color="#000000"
    />
     
    black
  </Styled(styled.div)>
</styled.div>
`;
